### ISSUE TITLE
Improve the ability to monitor related models for change

### ISSUE THREAD

**Issue by KeithWoods** [esp/esp-js#66]
Currently when a model is changed the router broadcasts a special event called `modelChangedEvent`. This allows an observer to react to the model and perhaps reflect some of that models state on their own.  

The problem with this is there is no good way to then stop the related models `modelChangedEvent` from being consumed by the initial model. Even if the initial model was actually subscribed to a `modelChangedEvent` for a completely unrelated model. For example:
Model A observes `modelChangedEvent` for model Z
Model B observes `modelChangedEvent` for model A
Note that when any model is change, for example an unrelated model C, both models A and B will receive an event and they have to filter it as it's not model Z or A retrospectively. However because they received the event, they are deemed 'changed' so the router will dispatch a `modelChangedEvent` for each and they'll just bounce off each other forever processing endless `modelChangedEvent`s. 

What's missing is the ability for the router to know which model you really want to observe when  the `modelChangedEvent` is fired but the api doesn't allow for it. 

i.e: this does not exist as the api has no event specific parameters (and nor should it).
```
_router.getEventStream('modelA', 'modelChagnedEvent', 'onlyModelZ')
```

You could perhaps solve it for the decorator observation by creating a new decorator just for ` modelChagnedEvent` however the underlying API still has the issue. 

We already have the model observation via `Router.getModelObservable` The only downside is you're not on the related models dispatch loop. For example:

```
// in modelA.js

_router.getModelObservable('modelZ').observe(modelZ => {
  // you can't update local state here 
  // as your're not on a dispatch loop for modelA
 //  you'd have to do this:
 _router.runAction(() => {
    // ok to update state here
  });
})

```

The `getModelObservable` above solution isn't ideal but works today. 

I think adding proper support to the `Router` to cater for related models getting changed rather than re-using the event publication API might be required. 

There are 2 options below. Note that regardless of the method signigure, both implementation under the hood will be similar. i.e. they will effectively add better built in support for `modelChangedEvent` by allowing this special event to be associated with the model the observer is interested in. This should enable the `Router` do selectively dispatch the `modelChangedEvent` rather than blindly dispatching and letting the observation handler filter.  With this approach the `Router` won't mark the model as changed as it never dispatched an event to it.

1 Create a new method on `Router`
```
router.getModelChangeEventStream(targetModel, relatedModel).observe(relatedModel => {
});
```

2 `allow events to be rich types`
```
router.getEventObservable(
  targetModel, { 
    eventName: 'modelChangedEvent'
    relatedModelId: 'relatedModelId'  
  },  ).observe(relatedModel => {
});
```

I think option 2 is more inline with the current API. 

Both options effectively change the underlying router api so the decorator observation can back onto either option.

**Comment by KeithWoods** [comment]
I made some changes for this in 91bc28fb756834eab1f591d9c36316553bb0be84. In effect I implemented option 2 and ensured that if you subscribe for a `modelChangedEvent` using a simple string you'd get an error throw.

Unfortunately after further thought there is still the possibility of ending up in an infinite loop for the same same scenario mentioned above. i.e. ModelA depends on ModelC, ModelB depends on ModelA and ModelC depends on ModelB. There is no direct circular dependency, but there is an indirect circular dependency. 

I think this could be solved by adding some reentrancy detection. The detection would be a simple event dispatch register whereby you denote you're dispatching `modelChangedEven`s, then proceed to process observers on dependent models and track their follow on `modelChangedEvent`s. If you end up dispatching to a model that you already dispatched to you simply break and throw. This should allow for a model dependency graph that matches a tree based hierarchical. Siblings could depends on one another (but not bi-directional) and children could depend upon their parents but not (bi directional).

The current changes in 91bc28fb756834eab1f591d9c36316553bb0be84 make slight improvements so can go into master, however I think this store can go on the back burner for now. Will fix timer permitting.
